File: python/taflow/indicators/rolling_average_drawdown.py
"""Public adapter for native rolling average drawdown."""

from typing import Any

import numpy as np

from .._native import RollingAverageDrawdown as _Native
from .._series import as_float64_series


class RollingAverageDrawdown:
    """Average maximum depth of distinct drawdown episodes in each window.

    An episode begins below a running peak and ends on recovery. Its depth is
    measured at its trough; the result averages episode depths rather than
    every underwater bar. ``period - 1`` leading outputs are ``NaN``. The
    independent name mapping is Wickra ``AverageDrawdown``.

    Args:
        values: Chronological equity-curve observations.
        period: Rolling window length, default 14.

    Raises:
        ValueError: If ``period`` is zero.
    """

    def __init__(self, period: int = 14) -> None:
        """Initialize an empty configured native state.
        """
        self._state = _Native(period)

    def append(self, value: float) -> "RollingAverageDrawdown":
        """Append one equity observation and return this instance."""
        self._state.append(float(value))
        return self

    def extend(self, values: Any) -> "RollingAverageDrawdown":
        """Append a chronological float64-compatible series and return self."""
        self._state.extend(as_float64_series(values))
        return self

    @property
    def value(self) -> float | None:
        """Return the latest average drawdown, or ``None`` during warm-up."""
        return self._state.value

    def compute(self) -> np.ndarray:
        """Return aligned history with warm-up represented by ``NaN``."""
        return self._state.compute()

    def reset(self) -> "RollingAverageDrawdown":
        """Clear the rolling state and return this instance."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of observations stored by native state."""
        return len(self._state)

File: crates/taflow-core/src/indicators/rolling_average_drawdown.rs
use crate::error::TaResult;
use crate::indicators::rolling_statistic_helpers::RollingValues;
use crate::stream::StreamingIndicator;

/// Mean maximum depth of distinct drawdown episodes in a rolling window.
#[derive(Debug, Clone)]
pub struct RollingAverageDrawdown {
    values: RollingValues,
    value: Option<f64>,
}

impl RollingAverageDrawdown {
    /// Create a rolling average-drawdown state with a non-zero `period`.
    pub fn new(period: usize) -> TaResult<Self> {
        Ok(Self {
            values: RollingValues::new(period)?,
            value: None,
        })
    }

    /// Append one equity-curve observation and return the latest result.
    pub fn append(&mut self, input: f64) -> Option<f64> {
        self.values.push(input);
        self.value = self.values.is_full().then(|| {
            let mut peak = f64::NEG_INFINITY;
            let mut episode_peak = 0.0_f64;
            let mut episode_trough = 0.0_f64;
            let mut depth_sum = 0.0;
            let mut episodes = 0_u32;
            let mut in_drawdown = false;

            for &equity in self.values.iter() {
                if equity >= peak {
                    if in_drawdown {
                        if episode_peak > 0.0 {
                            depth_sum += (episode_peak - episode_trough) / episode_peak;
                            episodes += 1;
                        }
                        in_drawdown = false;
                    }
                    peak = equity;
                } else if in_drawdown {
                    episode_trough = episode_trough.min(equity);
                } else {
                    in_drawdown = true;
                    episode_peak = peak;
                    episode_trough = equity;
                }
            }
            if in_drawdown && episode_peak > 0.0 {
                depth_sum += (episode_peak - episode_trough) / episode_peak;
                episodes += 1;
            }
            if episodes == 0 {
                0.0
            } else {
                depth_sum / f64::from(episodes)
            }
        });
        self.value
    }

    /// Return the latest value, or `None` during warm-up.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Restore fresh-state behavior while retaining allocated buffers.
    pub fn reset(&mut self) {
        self.values.clear();
        self.value = None;
    }
}

impl StreamingIndicator for RollingAverageDrawdown {
    type Output = f64;
    fn append(&mut self, input: f64) -> Option<f64> {
        Self::append(self, input)
    }
    fn value(&self) -> Option<f64> {
        self.value
    }
    fn reset(&mut self) {
        Self::reset(self)
    }
}
